File: algorithms/_test/paradigmas-greedy.py
# Greedy

# # Interval Scheduling
# 1. Interval Scheduling (máximo de atividades não sobrepostas)
def interval_scheduling(intervals):
    """
    intervals: lista de tuplas (start, end)
    Retorna o subconjunto máximo de intervalos que não se sobrepõem.
    """
    # ordena por tempo de término
    sorted_intervals = sorted(intervals, key=lambda x: x[1])
    result = []
    last_end = -float('inf')
    for start, end in sorted_intervals:
        if start >= last_end:
            result.append((start, end))
            last_end = end
    return result

if __name__ == "__main__":
    activities = [(1,4), (3,5), (0,6), (5,7), (3,9), (5,9), (6,10), (8,11), (8,12), (2,14), (12,16)]
    chosen = interval_scheduling(activities)
    print("Atividades escolhidas:", chosen)
    # saída esperada:
    # Atividades escolhidas: [(1,4), (5,7), (8,11), (12,16)]


# # Troco com moedas
# 2. Troco com moedas (mínimo de moedas via programação dinâmica)
def min_coins(coins, amount):
    """
    coins: lista de valores de moedas (inteiros positivos)
    amount: alvo
    Retorna o número mínimo de moedas para compor `amount`, ou -1 se não for possível.
    """
    INF = amount + 1
    dp = [0] + [INF] * amount
    for i in range(1, amount + 1):
        for c in coins:
            if c <= i and dp[i - c] + 1 < dp[i]:
                dp[i] = dp[i - c] + 1
    return dp[amount] if dp[amount] != INF else -1

if __name__ == "__main__":
    moedas = [1, 3, 4]
    valor = 6
    print("Min moedas para", valor, ":", min_coins(moedas, valor))
    # saída esperada: Min moedas para 6 : 2  (por exemplo, 3+3 ou 4+1+1)


# # Huffman Coding
# 3. Huffman Coding (construção de árvore sem heapq)
class Node:
    def __init__(self, char=None, freq=0, left=None, right=None):
        self.char  = char
        self.freq  = freq
        self.left  = left
        self.right = right

def build_huffman(freqs):
    """
    freqs: dicionário {char: frequência}
    Retorna raiz da árvore de Huffman.
    """
    # cria nós iniciais
    nodes = [Node(c, f) for c, f in freqs.items()]
    # itera até sobrar um nó
    while len(nodes) > 1:
        # encontra os dois nós de menor frequência
        nodes.sort(key=lambda n: n.freq)
        a = nodes.pop(0)
        b = nodes.pop(0)
        # combina em novo nó
        merged = Node(None, a.freq + b.freq, a, b)
        nodes.append(merged)
    return nodes[0]

def make_codes(root):
    """
    root: raiz da árvore de Huffman
    Retorna dicionário {char: código em bits (string)}.
    """
    codes = {}
    def dfs(node, prefix):
        if node is None:
            return
        # folha
        if node.char is not None:
            codes[node.char] = prefix or "0"
        else:
            dfs(node.left,  prefix + "0")
            dfs(node.right, prefix + "1")
    dfs(root, "")
    return codes

if __name__ == "__main__":
    freq_map = {'a': 5, 'b': 9, 'c': 12, 'd': 13, 'e': 16, 'f': 45}
    root = build_huffman(freq_map)
    huff_codes = make_codes(root)
    print("Códigos de Huffman:")
    for ch, code in huff_codes.items():
        print(f" {ch}: {code}")
    # exemplo de saída (códigos podem variar em estrutura):
    #  f: 0
    #  c: 100
    #  d: 101
    #  a: 1100
    #  b: 1101
    #  e: 111

File: algorithms/_test/paradigmas-greedy.rb
# Greedy Algorithms em Ruby

# 1. Interval Scheduling (máximo de atividades não sobrepostas)
def interval_scheduling(intervals)
  # ordena por tempo de término
  sorted = intervals.sort_by { |start, finish| finish }
  result = []
  last_end = -Float::INFINITY

  sorted.each do |start, finish|
    if start >= last_end
      result << [start, finish]
      last_end = finish
    end
  end

  result
end

# 2. Troco com moedas (mínimo de moedas via DP)
def min_coins(coins, amount)
  inf = amount + 1
  dp = [0] + [inf] * amount

  (1..amount).each do |i|
    coins.each do |c|
      if c <= i && dp[i - c] + 1 < dp[i]
        dp[i] = dp[i - c] + 1
      end
    end
  end

  dp[amount] == inf ? -1 : dp[amount]
end

# 3. Huffman Coding (construção de árvore sem heap)
class Node
  attr_accessor :char, :freq, :left, :right
  def initialize(char = nil, freq = 0, left = nil, right = nil)
    @char, @freq, @left, @right = char, freq, left, right
  end
end

def build_huffman(freqs)
  nodes = freqs.map { |c, f| Node.new(c, f) }
  while nodes.size > 1
    nodes.sort_by!(&:freq)
    a = nodes.shift
    b = nodes.shift
    merged = Node.new(nil, a.freq + b.freq, a, b)
    nodes << merged
  end
  nodes.first
end

def make_codes(root)
  codes = {}
  dfs = lambda do |node, prefix|
    return if node.nil?
    if node.char
      codes[node.char] = prefix.empty? ? "0" : prefix
    else
      dfs.call(node.left,  prefix + "0")
      dfs.call(node.right, prefix + "1")
    end
  end
  dfs.call(root, "")
  codes
end

# Exemplos de uso
if __FILE__ == $PROGRAM_NAME
  puts "--- Interval Scheduling ---"
  activities = [
    [1,4], [3,5], [0,6], [5,7], [3,9],
    [5,9], [6,10], [8,11], [8,12], [2,14], [12,16]
  ]
  chosen = interval_scheduling(activities)
  p "Atividades escolhidas: #{chosen}"
  # => Atividades escolhidas: [[1,4], [5,7], [8,11], [12,16]]

  puts "\n--- Troco com Moedas ---"
  coins = [1, 3, 4]
  amount = 6
  p "Min moedas para #{amount}: #{min_coins(coins, amount)}"
  # => Min moedas para 6: 2

  puts "\n--- Huffman Coding ---"
  freq_map = { 'a' => 5, 'b' => 9, 'c' => 12,
               'd' => 13, 'e' => 16, 'f' => 45 }
  root = build_huffman(freq_map)
  codes = make_codes(root)
  puts "Códigos de Huffman:"
  codes.each { |ch, code| puts " #{ch}: #{code}" }
end
